traitsui/examples/demo/Advanced/Dynamically_changing_buttons_demo.py

```diff
@@ -0,0 +1,86 @@
+# (C) Copyright 2004-2021 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+"""
+This demo shows how to dynamically change the label or image of a button. Note
+in this demo clicking the button itself does nothing.
+
+Please refer to the `ButtonEditor API docs`_ for further information.
+
+.. _ButtonEditor API docs: https://docs.enthought.com/traitsui/api/traitsui.editors.button_editor.html#traitsui.editors.button_editor.ButtonEditor
+"""
+from pyface.api import ImageResource
+from traits.api import Button, Enum, HasTraits, Instance, List, Str
+
+from traitsui.api import (
+    ButtonEditor,
+    Group,
+    ImageEditor,
+    InstanceChoice,
+    InstanceEditor,
+    Item,
+    UItem,
+    View
+) 
+
+
+class ImageChoice(InstanceChoice):
+    def get_view(self):
+        return View(
+            UItem('name', editor=ImageEditor(image=self.object))
+        )
```

Comment:
I went with the InstanceEditor approach because trying to use `ImageEnumEditor` was giving me trouble/ I couldn't get things working how I wanted them to.  
This demo is now a little bit showcasing the _secret_ InstanceEditor Api as well.

Also, relating to my other comment, it seems that if I had used the `Image` trait correctly here, this code would be un needed as the Image trait sets the editor to ImageEditor for the ImageResource that is stored.